File: backends/rust-axum/src/infrastructure/persistence/staff.rs
//! スタッフリポジトリ SQLx 実装モジュール。
//!
//! # Author
//! Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use sqlx::{MySqlPool, QueryBuilder};
use crate::domain::staff::{
    condition::Condition,
    entity::Staff,
    repository::{DomainError, Repository},
};

#[derive(sqlx::FromRow)]
struct StaffRow {
    id:            u32,
    name:          String,
    email:         String,
    provider:      i32,
    provider_id:   String,
    avatar:        Option<String>,
    role:          u32,
    last_login_at: Option<DateTime<Utc>>,
    created_at: DateTime<Utc>,
    created_by:    Option<u32>,
    updated_at: DateTime<Utc>,
    updated_by:    Option<u32>,
    deleted_at: Option<DateTime<Utc>>,
    deleted_by:    Option<u32>,
    version:       u32,
}

fn row_to_entity(r: StaffRow) -> Staff {
    Staff {
        id:            r.id,
        name:          r.name,
        email:         r.email,
        provider:      r.provider,
        provider_id:   r.provider_id,
        avatar:        r.avatar,
        role:          r.role as i32,
        last_login_at: r.last_login_at,
        created_at:    r.created_at,
        created_by:    r.created_by,
        updated_at:    r.updated_at,
        updated_by:    r.updated_by,
        deleted_at:    r.deleted_at,
        deleted_by:    r.deleted_by,
        version:       r.version as i32,
    }
}

const ALLOWED_SORT: &[&str] = &["name", "role", "created_at"];

/// SQLx を用いたスタッフリポジトリ実装。
pub struct SqlxStaffRepository {
    pool: MySqlPool,
}

impl SqlxStaffRepository {
    /// プールを受け取りリポジトリを生成します。
    pub fn new(pool: MySqlPool) -> Self {
        Self { pool }
    }

    async fn fetch_by_id(&self, id: u32) -> Result<Option<Staff>, DomainError> {
        let row = sqlx::query_as::<_, StaffRow>(
            "SELECT * FROM staffs WHERE id = ?"
        )
        .bind(id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(row_to_entity))
    }

    fn push_filters<'a>(qb: &mut QueryBuilder<'a, sqlx::MySql>, cond: &'a Condition) {
        if let Some(kw) = &cond.keyword {
            if !kw.is_empty() {
                let like = format!("%{}%", kw);
                qb.push(" AND (name LIKE ");
                qb.push_bind(like.clone());
                qb.push(" OR email LIKE ");
                qb.push_bind(like);
                qb.push(")");
            }
        }
        if !cond.roles.is_empty() {
            qb.push(" AND role IN (");
            let mut sep = qb.separated(", ");
            for r in &cond.roles {
                sep.push_bind(*r);
            }
            qb.push(")");
        }
    }
}

#[async_trait]
impl Repository for SqlxStaffRepository {
    async fn count_by_condition(&self, cond: Condition) -> Result<i64, DomainError> {
        let mut qb: QueryBuilder<sqlx::MySql> = QueryBuilder::new(
            "SELECT COUNT(*) FROM staffs WHERE 1=1"
        );
        SqlxStaffRepository::push_filters(&mut qb, &cond);
        let row: (i64,) = qb.build_query_as().fetch_one(&self.pool).await?;
        Ok(row.0)
    }

    async fn find_by_condition(&self, cond: Condition) -> Result<Vec<Staff>, DomainError> {
        let mut qb: QueryBuilder<sqlx::MySql> = QueryBuilder::new(
            "SELECT * FROM staffs WHERE 1=1"
        );
        SqlxStaffRepository::push_filters(&mut qb, &cond);
        let sort_col = match cond.sort.as_deref() {
            Some(s) if ALLOWED_SORT.contains(&s) => s,
            _ => "id",
        };
        let sort_dir = if cond.sort_type.as_deref() == Some("desc") { "DESC" } else { "ASC" };
        qb.push(format!(" ORDER BY {} {}", sort_col, sort_dir));
        let limit = if cond.limit > 0 { cond.limit } else { 10 };
        qb.push(" LIMIT ");
        qb.push_bind(limit);
        qb.push(" OFFSET ");
        qb.push_bind(cond.offset);
        let rows = qb.build_query_as::<StaffRow>().fetch_all(&self.pool).await?;
        Ok(rows.into_iter().map(row_to_entity).collect())
    }

    async fn find_by_id(&self, id: u32) -> Result<Option<Staff>, DomainError> {
        let row = sqlx::query_as::<_, StaffRow>(
            "SELECT * FROM staffs WHERE id = ? AND deleted_at IS NULL LIMIT 1"
        )
        .bind(id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(row_to_entity))
    }

    async fn find_by_provider(&self, provider: i32, provider_id: &str) -> Result<Option<Staff>, DomainError> {
        let row = sqlx::query_as::<_, StaffRow>(
            "SELECT * FROM staffs WHERE provider = ? AND provider_id = ? LIMIT 1"
        )
        .bind(provider)
        .bind(provider_id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(row_to_entity))
    }

    async fn find_all_active(&self) -> Result<Vec<Staff>, DomainError> {
        let rows = sqlx::query_as::<_, StaffRow>(
            "SELECT * FROM staffs WHERE deleted_at IS NULL ORDER BY id ASC"
        )
        .fetch_all(&self.pool)
        .await?;
        Ok(rows.into_iter().map(row_to_entity).collect())
    }

    async fn save(&self, s: Staff) -> Result<Staff, DomainError> {
        if s.id == 0 {
            let result = sqlx::query(
                "INSERT INTO staffs (name, email, provider, provider_id, avatar, role, \
                 last_login_at, created_at, created_by, updated_at, updated_by, \
                 deleted_at, deleted_by, version) \
                 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
            )
            .bind(&s.name)
            .bind(&s.email)
            .bind(s.provider)
            .bind(&s.provider_id)
            .bind(&s.avatar)
            .bind(s.role)
            .bind(s.last_login_at)
            .bind(s.created_at)
            .bind(s.created_by)
            .bind(s.updated_at)
            .bind(s.updated_by)
            .bind(s.deleted_at)
            .bind(s.deleted_by)
            .bind(s.version)
            .execute(&self.pool)
            .await?;
            let new_id = result.last_insert_id() as u32;
            Ok(self.fetch_by_id(new_id).await?.unwrap())
        } else {
            let result = sqlx::query(
                "UPDATE staffs SET name = ?, email = ?, provider = ?, provider_id = ?, avatar = ?, \
                 role = ?, last_login_at = ?, updated_at = ?, updated_by = ?, \
                 deleted_at = ?, deleted_by = ?, version = version + 1 \
                 WHERE id = ? AND version = ?"
            )
            .bind(&s.name)
            .bind(&s.email)
            .bind(s.provider)
            .bind(&s.provider_id)
            .bind(&s.avatar)
            .bind(s.role)
            .bind(s.last_login_at)
            .bind(s.updated_at)
            .bind(s.updated_by)
            .bind(s.deleted_at)
            .bind(s.deleted_by)
            .bind(s.id)
            .bind(s.version)
            .execute(&self.pool)
            .await?;
            if result.rows_affected() == 0 {
                return Err("optimistic_lock_conflict".to_string().into());
            }
            Ok(self.fetch_by_id(s.id).await?.unwrap())
        }
    }

    async fn update_role(&self, id: u32, role: i32, updated_by: u32, version: i32) -> Result<bool, DomainError> {
        let now = chrono::Utc::now();
        let result = sqlx::query(
            "UPDATE staffs SET role = ?, updated_at = ?, updated_by = ?, version = version + 1 \
             WHERE id = ? AND deleted_at IS NULL AND version = ?"
        )
        .bind(role)
        .bind(now)
        .bind(updated_by)
        .bind(id)
        .bind(version)
        .execute(&self.pool)
        .await?;
        if result.rows_affected() == 0 {
            return Err("optimistic_lock_conflict".to_string().into());
        }
        Ok(true)
    }

    async fn soft_delete(&self, id: u32, deleted_by: u32, version: i32) -> Result<bool, DomainError> {
        let now = chrono::Utc::now();
        let result = sqlx::query(
            "UPDATE staffs SET deleted_at = ?, deleted_by = ? \
             WHERE id = ? AND deleted_at IS NULL AND version = ?"
        )
        .bind(now)
        .bind(deleted_by)
        .bind(id)
        .bind(version)
        .execute(&self.pool)
        .await?;
        if result.rows_affected() == 0 {
            return Err("optimistic_lock_conflict".to_string().into());
        }
        Ok(true)
    }

    async fn restore(&self, id: u32) -> Result<bool, DomainError> {
        let result = sqlx::query(
            "UPDATE staffs SET deleted_at = NULL, deleted_by = NULL \
             WHERE id = ? AND deleted_at IS NOT NULL"
        )
        .bind(id)
        .execute(&self.pool)
        .await?;
        Ok(result.rows_affected() > 0)
    }
}

File: backends/csharp-aspnet/src/Authorization.Api/Domain/Staff/Staff.cs
/*
 * スタッフ ドメインモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
namespace Authorization.Api.Domain.Staff;

/// <summary>スタッフ権限コードです。</summary>
public static class StaffRole
{
    public const int Admin  = 1;
    public const int Member = 2;

    /// <summary>不正な値は Member に丸めます。</summary>
    /// <param name="value">ロールコード</param>
    /// <returns>ロールコード。Admin以外はMemberに丸める</returns>
    public static int From(int value) => value == Admin ? Admin : Member;

    /// <summary>有効な権限コードかどうか。</summary>
    /// <param name="value">ロールコード</param>
    /// <returns>Admin/Memberのいずれかであればtrue</returns>
    public static bool IsValid(int value) => value is Admin or Member;
}

/// <summary>OAuth プロバイダーコードです。</summary>
public static class StaffProvider
{
    public const int Google = 1;
    public const int Github = 2;
}

/// <summary>スタッフエンティティです。</summary>
public sealed record Staff
{
    public long      Id          { get; init; }
    public string    Name        { get; init; } = "";
    public string    Email       { get; init; } = "";
    public int       Provider    { get; init; }
    public string    ProviderId  { get; init; } = "";
    public string?   Avatar      { get; init; }
    public int       Role        { get; init; } = StaffRole.Member;
    public DateTime? LastLoginAt { get; init; }
    public DateTime  CreatedAt   { get; init; }
    public long?     CreatedBy   { get; init; }
    public DateTime  UpdatedAt   { get; init; }
    public long?     UpdatedBy   { get; init; }
    public DateTime? DeletedAt   { get; init; }
    public long?     DeletedBy   { get; init; }
    public int       Version     { get; init; } = 1;

    /// <summary>状態コード（1=有効、0=無効）。deleted_at の有無で算出します。</summary>
    public int Status => DeletedAt is null ? 1 : 0;
}

/// <summary>スタッフ検索条件です。</summary>
public sealed record StaffCondition
{
    public string?   Keyword  { get; init; }
    public List<int> Roles    { get; init; } = [];
    public int       Offset   { get; init; }
    public int       Limit    { get; init; } = 10;
    public string?   Sort     { get; init; }
    public string?   SortType { get; init; }
}

/// <summary>一覧表示用のスタッフ VO です。</summary>
public sealed record StaffListItem(
    long Id,
    string Name,
    string Email,
    int Role,
    int Status,
    DateTime CreatedAt,
    DateTime UpdatedAt,
    int Version
);

/// <summary>ログイン中スタッフのプロフィール VO です。</summary>
public sealed record StaffVo(long Id, string Name, string? Avatar, int Role);

/// <summary>スタッフリポジトリです。</summary>
public interface IStaffRepository
{
    /// <summary>条件に一致するスタッフの総件数を返します（ページング無視）。</summary>
    /// <param name="cond">検索条件</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>総件数</returns>
    Task<int> CountByConditionAsync(StaffCondition cond, CancellationToken ct = default);

    /// <summary>条件に一致するスタッフ一覧を返します。</summary>
    /// <param name="cond">検索条件</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>スタッフ一覧（ページング適用済み）</returns>
    Task<List<Staff>> FindByConditionAsync(StaffCondition cond, CancellationToken ct = default);

    /// <summary>IDでスタッフを取得します。</summary>
    /// <param name="id">スタッフID</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>スタッフ、存在しない場合はnull</returns>
    Task<Staff?> FindByIdAsync(long id, CancellationToken ct = default);

    /// <summary>OAuthプロバイダー情報でスタッフを取得します。</summary>
    /// <param name="provider">OAuthプロバイダーコード</param>
    /// <param name="providerId">プロバイダー側のユーザーID</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>スタッフ、存在しない場合はnull</returns>
    Task<Staff?> FindByProviderAsync(int provider, string providerId, CancellationToken ct = default);

    /// <summary>論理削除されていない全スタッフを返します。</summary>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>有効なスタッフ一覧</returns>
    Task<List<Staff>> FindAllActiveAsync(CancellationToken ct = default);

    /// <summary>
    /// 保存します。Id が 0 なら新規登録、それ以外は楽観排他ロック付き更新です。
    /// </summary>
    /// <param name="staff">保存するスタッフ</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>保存後のスタッフ（IDやバージョンが反映済み）</returns>
    /// <exception cref="Support.AppException">バージョン不一致（409）</exception>
    Task<Staff> SaveAsync(Staff staff, CancellationToken ct = default);

    /// <summary>権限を更新します。</summary>
    /// <param name="id">スタッフID</param>
    /// <param name="role">新しい権限コード</param>
    /// <param name="updatedBy">操作を実行したスタッフID</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>更新できた場合は true</returns>
    Task<bool> UpdateRoleAsync(long id, int role, long updatedBy, CancellationToken ct = default);

    /// <summary>論理削除します。</summary>
    /// <param name="id">スタッフID</param>
    /// <param name="deletedBy">削除を実行したスタッフID</param>
    /// <param name="version">楽観排他ロック用バージョン番号</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>削除できた場合は true</returns>
    /// <exception cref="Support.AppException">バージョン不一致（409）</exception>
    Task<bool> SoftDeleteAsync(long id, long deletedBy, int version, CancellationToken ct = default);

    /// <summary>論理削除を取り消します。</summary>
    /// <param name="id">スタッフID</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>復元できた場合は true</returns>
    Task<bool> RestoreAsync(long id, CancellationToken ct = default);
}
